Track shard snapshot progress during node shutdown

DiannaHohensee opened the pull request:
Track shard snapshot progress during shutdown to identify any
bottlenecks that cause slowness that can ultimately block shard
re-allocation.

Relates ES-9086

------------

In a follow up, I want to add
- human readable time since shutdown received
- time it took to set PAUSING
- time it took to finish stopping all the shard snapshots, to completely unblock shard re-allocation.

Among other things. But better to ship it sooner.

DiannaHohensee committed:
Track shard snapshot progress during node shutdown

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @DiannaHohensee, I've created a changelog YAML for you.

DiannaHohensee committed:
Update docs/changelog/112567.yaml

ywangd reviewed (commented):
I left some low level comments. I also have a higher level comment about what we want to track and how effective the current approach is.

IIUC, the problem we want to investigate is why some shard snapshots keep running or new shard snapshots starting when a node is marked for shutdown. So as a first step, we want to know more details about the progress of shard snapshots on a node. Rather than tracking and logging the individual actions on starting, pausing or sending/receving requests, I wonder whether we should log the actual shardSnapshotStatus which is already tracked as [SnapshotsShardsService#shardSnapshots](https://github.com/elastic/elasticsearch/blob/8e9e6532fead9e4debda8df8a18d09fb1fada899/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java#L85) and exposed via a [method](https://github.com/elastic/elasticsearch/blob/8e9e6532fead9e4debda8df8a18d09fb1fada899/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java#L194). Ideally this field should be either empty or have all its entries in an final stage, such as `PAUSED`, at some point after shutdown begins. If a shard snapshot is in the progress of being pausing, it should be in the `PAUSING` stage. What I have in mind is something like the follows:
1. At the beginning of shutdown, right after we call `handleUpdatedSnapshotsInProgressEntry` for each shard, we can go through this field and log shards that are not in the desired states such as `PAUSING`, `PAUSED`, `FAILED` etc. In theory, every shard should be in such state.
2. Periodically go through the field again and check the progress, e.g. how many `PAUSING` shards left and maybe log the top 5 or 10 of them. 
3. Once every shard is in a final stage, log an message to say the progress is completed.
4. If any new shard snapshot starts during shutdown, immediately log a message since this should _not_ happen.
5. If there is any master fail over, we can log a similar message for this field again and with a notion to say this is from master fail over.

Compare to counting individual actions, I think above potentially gives us more details such as index name and shard id and potentially transitations between different stages. Counting may also have an issue with master fail over, e.g. if the same request is sent to both old and new master and both of them send a response, the counting will be off for sometime. Tracking requests/response feels a bit indirect since what we are really interested is the stage of each shard snapshot.

Thanks for reading through the long comment.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -129,11 +134,38 @@ protected void doClose() {
         try {
+            final var localNodeId = clusterService.localNode().getId();
+
+            // Track when this node enters and leaves shutdown mode because we pause shard snapshots for shutdown.
+            // The snapshotShutdownProgressTracker will report (via logging) on the progress shard snapshots make
+            // towards either completing (successfully or otherwise) or pausing.
+            NodesShutdownMetadata currentShutdownMetadata = event.state().metadata().custom(NodesShutdownMetadata.TYPE);
+            NodesShutdownMetadata previousShutdownMetadata = event.previousState().metadata().custom(NodesShutdownMetadata.TYPE);
+            SingleNodeShutdownMetadata currentLocalNodeShutdownMetadata = currentShutdownMetadata != null
+                ? currentShutdownMetadata.get(localNodeId)
+                : null;
+            SingleNodeShutdownMetadata previousLocalNodeShutdownMetadata = previousShutdownMetadata != null
+                ? previousShutdownMetadata.get(localNodeId)
+                : null;
+
+            boolean isLocalNodeAddingShutdown = false;
+            if (previousLocalNodeShutdownMetadata == null && currentLocalNodeShutdownMetadata != null) {
+                snapshotShutdownProgressTracker.onClusterStateAddShutdown();
+                isLocalNodeAddingShutdown = true;
+            }
+
+            if (currentLocalNodeShutdownMetadata == null && previousLocalNodeShutdownMetadata != null) {
+                snapshotShutdownProgressTracker.onClusterStateRemoveShutdown();
+            }
```
I originally used SnapshotsInProgress#isNodeIdForRemoval, but it turned out to be unpredictable because that field gets cleared without cluster state changes -- IIRC, when the snapshots finish pausing, or something. I pivoted to looking at the shutdown marker as more predictable. But in theory, I think I could use isNodeIdForRemoval, I think 🤔  I shall think more on this.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -129,11 +134,38 @@ protected void doClose() {
+                : null;
+
+            boolean isLocalNodeAddingShutdown = false;
+            if (previousLocalNodeShutdownMetadata == null && currentLocalNodeShutdownMetadata != null) {
+                snapshotShutdownProgressTracker.onClusterStateAddShutdown();
+                isLocalNodeAddingShutdown = true;
```
Fair. I should add some testing around not turning on the tracker in this scenario. Thanks

DiannaHohensee commented:
> IIUC, the problem we want to investigate is why some shard snapshots keep running or new shard snapshots starting when a node is marked for shutdown. So as a first step, we want to know more details about the progress of shard snapshots on a node. 

Yes 👍 This first commit is about establishing a shard snapshot shutdown Tracker class, and getting some basic information to identify what phases the shard snapshots are stuck in, and then we can follow up as needed by surfacing additional pertinent information to drill down into the problem.

> Rather than tracking and logging the individual actions on starting, pausing or sending/receving requests, I wonder whether we should log the actual shardSnapshotStatus which is already tracked as [SnapshotsShardsService#shardSnapshots](https://github.com/elastic/elasticsearch/blob/8e9e6532fead9e4debda8df8a18d09fb1fada899/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java#L85) and exposed via a [method](https://github.com/elastic/elasticsearch/blob/8e9e6532fead9e4debda8df8a18d09fb1fada899/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java#L194). Ideally this field should be either empty or have all its entries in a final stage, such as `PAUSED`, at some point after shutdown begins. If a shard snapshot is in the progress of being pausing, it should be in the `PAUSING` stage. 

Essentially the information we would publish about a shard snapshot is this [IndexShardSnapshotStatus::toString](https://github.com/elastic/elasticsearch/blob/163b2c781f980ed0b3383f9f3b27b6e4d0932e6b/server/src/main/java/org/elasticsearch/index/snapshots/IndexShardSnapshotStatus.java#L405-L430) information. It would be nice to see that info 🤔 I wonder, though, what does that give us beyond seeing that a shard snapshot is stalling somewhere in the upload phase? The counter `numberOfShardSnapshotsInProgressOnDataNode` that I've added will inform us that upload is getting stuck, somewhere in that phase. We won't be seeing these logs live, most likely, we'll see them after an incident. The allocation code currently does this during shutdown, picking some random shard to log about being blocked on snapshotting. 

> What I have in mind is something like the follows:
1. At the beginning of shutdown, right after we call `handleUpdatedSnapshotsInProgressEntry` for each shard, we can go through this field and log shards that are not in the desired states such as `PAUSING`, `PAUSED`, `FAILED` etc. In theory, every shard should be in such state.
2. Periodically go through the field again and check the progress, e.g. how many `PAUSING` shards left and maybe log the top 5 or 10 of them. 
3. Once every shard is in a final stage, log an message to say the progress is completed.
4. If any new shard snapshot starts during shutdown, immediately log a message since this should _not_ happen.
5. If there is any master fail over, we can log a similar message for this field again and with a notion to say this is from master fail over.

> Compare to counting individual actions, I think above potentially gives us more details such as index name and shard id and potentially transitations between different stages. 

I like the idea of tracking master failure in some way. My concern about logging the IndexShardSnapshotStatus is that
1) it will produce potentially a great deal of logging if a snapshot of lots of shards is in progress at shutdown.
2) Logging a small subset of 5 or 10 randomly won't give us the necessary consistency to track periodic progress.
3) what does the information of file uploads, etc., give us compared to counters? What would we then do with that information? Again, I would expect us to see the logs after the incident, not during.

> Counting may also have an issue with master fail over, e.g. if the same request is sent to both old and new master and both of them send a response, the counting will be off for sometime. Tracking requests/response feels a bit indirect since what we are really interested is the stage of each shard snapshot.

When I added the master request counter hooks to [sendSnapshotShardUpdate()](https://github.com/elastic/elasticsearch/blob/163b2c781f980ed0b3383f9f3b27b6e4d0932e6b/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java#L667-L692), I thought that the retries and master failover handling would be handled internally, and the places I hook into were the start and finish points regardless of those activities. Sounds like I should add a master failover test to find out.

---------

In a bit of haste to turn this around before my half day today ends (I'm out-of-office Friday), so you can see it tomorrow before your vacation. Leaving the code changes aside for now. Definitely point out if I overlooked something in my haste :) And I much appreciate you taking an in-depth look.

DiannaHohensee reviewed (commented):
Responding to some comments. Leaving aside the clearer code change related ones for now, just to turn this around, but they look good and I'll address them 👍

DiannaHohensee committed:
Merge branch 'main' into 2024/08/05/track-shard-snapshot-progress-tested

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+    private void scheduleProgressLogger() {
+        scheduledProgressLoggerFuture = threadPool.scheduleWithFixedDelay(
+            progressLogger,
+            progressLoggerInterval,
```
Should we special-case `TimeValue#MINUS_ONE` to mean to disable this feature? At the moment I think that would mean to emit progress logs constantly with zero delay.

DaveCTurner replied:
Yeah I think so. `1ms` is clearly going to cause a tight loop but I think that would be surprising behaviour for `-1` (and arguably for `0` too).

The advantage of having a value which completely disables this functionality is that it gives a potential workaround for any bugs that might crop up later, in a way that running it daily would not.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    /** How frequently shard snapshot progress is logged after receiving local node shutdown metadata. */
+    public static final Setting<TimeValue> SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING = Setting.positiveTimeSetting(
```
WDYT about making this setting dynamic? I don't think that would add much complexity.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+        this.shutdownStartMillis = -1;
+        this.shutdownFinishedSignallingPausingMillis = -1;
+
+        // Turn off the progress logger, which we only want to run during shutdown.
+        cancelProgressLogger();
```
I think it will stop naturally in that case anyway when the threadpools shut down. There's no need to stop it sooner is there?

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -382,6 +423,7 @@ public void onResponse(ShardSnapshotResult shardSnapshotResult) {
                         snapshotStatus.generation()
                     );
                 }
+                snapshotShutdownProgressTracker.decNumberOfShardSnapshotsInProgress(snapshotStatus.getStage());
```
I'd rather we did this with an `ActionListener#runBefore` just to make it clearer to readers that it always happens.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -665,17 +708,20 @@ private void notifyUnsuccessfulSnapshotShard(
                 @Override
                 public void onResponse(Void aVoid) {
                     logger.trace("[{}][{}] updated snapshot state to [{}]", shardId, snapshot, status);
+                    snapshotShutdownProgressTracker.releaseRequestSentToMaster(snapshot, shardId);
```
Likewise here, could we use `runBefore` or `runAfter`?

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+     * Logs some statistics about shard snapshot progress.
+     */
+    private void logProgressReport() {
+        assert this.shutdownStartMillis != -1;
```
I think this might not always be true, if we're cancelled on another thread just before executing this line.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+     * @param shardId second part of a unique identifier
+     */
+    public void trackRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        logger.trace("Tracking shard snapshot request to master");
```
I reckon we should report the snapshot and shard in this message.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+     * @param shardId second part of a unique identifier
+     */
+    public void releaseRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        logger.trace("A shard snapshot request to master finished");
```
I reckon we should report the snapshot and shard in this message. It might also be worth reporting the elapsed time since we sent the message?

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+     * Tracks how many shard snapshots are started.
+     */
+    public void incNumberOfShardSnapshotsInProgress() {
+        logger.trace("New shard snapshot started");
```
I'd suggest reporting the shard and snapshot here too.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,257 @@
+     * Tracks how many shard snapshots have finished since shutdown mode began.
+     */
+    public void decNumberOfShardSnapshotsInProgress(IndexShardSnapshotStatus.Stage stage) {
+        logger.trace(Strings.format("A shard snapshot finished in state [%s]", stage));
```
I'd suggest reporting the shard and snapshot here too.

DaveCTurner reviewed (commented):
Overall looks good to me, I left a few comments in addition to Yang's

DiannaHohensee committed:
Merge branch 'main' into 2024/08/05/track-shard-snapshot-progress-tested

DiannaHohensee committed:
Changes per David's review

DiannaHohensee committed:
Merge branch 'main' into 2024/08/05/track-shard-snapshot-progress-tested

DiannaHohensee committed:
fix new file headers

DiannaHohensee reviewed (commented):
Looks like I've broken a couple miscellaneous tests, so I'll follow up on that. Otherwise I've turned around the review feedback

DaveCTurner commented on server/src/test/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTrackerTests.java:
```diff
@@ -0,0 +1,252 @@
+            TimeValue.timeValueMillis(200)
+        )
+        .build();
+    TestThreadPool testThreadPool;
```
Since this is the only thing scheduling any future tasks in the context of this test I think you should be able to just `advanceTime()` once and then `runAllRunnableTasks()` to trigger the next check. I would expect `runTasksUpToTimeInOrder()` to work too, not sure what problems you faced there. Note it takes the target "absolute" time, rather than the number of simulated milliseconds to allow to elapse?

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -1545,6 +1546,34 @@ protected final DocWriteResponse indexDoc(String index, String id, Object... sou
+     * Index documents until all the shards are at least bytesPerShard in size.
+     * @return ShardStats for the newly populated shards.
+     */
+    protected ShardStats[] populateIndexShards(final String indexName, long bytesPerShard) {
```
I suspect the one in `[DiskThresholdDeciderIT.java](https://github.com/elastic/elasticsearch/pull/112567/files#diff-e44f19e24fcf88d32bd88d5ed8c4f19b7eb5c01560fc20168784476143cef808)` needs to be exclusive still, or else there's a (slim) chance of getting a shard that exactly fits into the watermark.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,271 @@
+
+    public SnapshotShutdownProgressTracker(
+        Supplier<String> localNodeIdSupplier,
+        BiConsumer<Setting<TimeValue>, Consumer<TimeValue>> addSettingsUpdateConsumer,
```
Could we just pass in a `ClusterSettings`? They're easy to find or make in tests.

DaveCTurner commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java:
```diff
@@ -0,0 +1,271 @@
+        ThreadPool threadPool
+    ) {
+        this.getLocalNodeId = localNodeIdSupplier;
+        this.progressLoggerInterval = SNAPSHOT_PROGRESS_DURING_SHUTDOWN_LOG_INTERVAL_SETTING.get(settings);
```
Suggest using `org.elasticsearch.common.settings.AbstractScopedSettings#initializeAndWatch` to avoid having to  set the field explicitly here

DaveCTurner reviewed (commented):
couple more nits, I'd like to see the tests using `DeterministicTaskQueue` too still, maybe share where you got stuck and I'll dig deeper

DiannaHohensee committed:
use deterministic task queue

DiannaHohensee committed:
use cluster settings, add dynamic setting and testing thereof

DiannaHohensee committed:
Merge branch 'main' into 2024/08/05/track-shard-snapshot-progress-tested

DiannaHohensee reviewed (commented):
Comments addressed.

I'm currently baffled as to why the NodeShutdownIT test has failures (doesn't reproduce locally, of course), since none of my changes should affect existing behavior.

DiannaHohensee committed:
Merge branch 'main' into 2024/08/05/track-shard-snapshot-progress-tested

DaveCTurner reviewed (approved):
LGTM

DaveCTurner commented:
The part-5 test failure looks germane tho, it's tripping an assertion here on the master:

```
[2024-10-08T15:31:16,488][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [javaRestTest-3] fatal error in thread [elasticsearch[javaRestTest-3][clusterApplierService#updateTask][T#1]], exiting
java.lang.AssertionError: Expected a call to add shutdown mode before a call to remove shutdown mode.
        at org.elasticsearch.snapshots.SnapshotShutdownProgressTracker.onClusterStateRemoveShutdown(SnapshotShutdownProgressTracker.java:185) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.snapshots.SnapshotShardsService.clusterChanged(SnapshotShardsService.java:167) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:566) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:552) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:511) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:433) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:157) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:925) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:218) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:184) ~[elasticsearch-9.0.0-SNAPSHOT.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1570) ~[?:?]
```

DaveCTurner commented:
... because I think we need to consider the type of both `previousLocalNodeShutdownMetadata` and `currentLocalNodeShutdownMetadata` when deciding whether we're starting or stopping a shutdown of this node. Something like this perhaps?

```diff
diff --git a/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java b/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
index 8c3401276e3..ca48e4c50e4 100644
--- a/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java
@@ -154,18 +154,13 @@ public final class SnapshotShardsService extends AbstractLifecycleComponent impl
                 : null;

             boolean isLocalNodeAddingShutdown = false;
-            if (previousLocalNodeShutdownMetadata == null
-                && currentLocalNodeShutdownMetadata != null
-                && currentLocalNodeShutdownMetadata.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+            if (isTrackedShutdown(previousLocalNodeShutdownMetadata) == false && isTrackedShutdown(currentLocalNodeShutdownMetadata)) {
                 snapshotShutdownProgressTracker.onClusterStateAddShutdown();
                 isLocalNodeAddingShutdown = true;
-            }
-
-            if (currentLocalNodeShutdownMetadata == null
-                && previousLocalNodeShutdownMetadata != null
-                && previousLocalNodeShutdownMetadata.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
-                snapshotShutdownProgressTracker.onClusterStateRemoveShutdown();
-            }
+            } else if (isTrackedShutdown(previousLocalNodeShutdownMetadata)
+                && isTrackedShutdown(currentLocalNodeShutdownMetadata) == false) {
+                    snapshotShutdownProgressTracker.onClusterStateRemoveShutdown();
+                }

             final var currentSnapshots = SnapshotsInProgress.get(event.state());

@@ -209,6 +204,10 @@ public final class SnapshotShardsService extends AbstractLifecycleComponent impl
         }
     }

+    private static boolean isTrackedShutdown(@Nullable SingleNodeShutdownMetadata localNodeShutdownMetadata) {
+        return localNodeShutdownMetadata != null && localNodeShutdownMetadata.getType() != SingleNodeShutdownMetadata.Type.RESTART;
+    }
+
     @Override
     public void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {
         // abort any snapshots occurring on the soon-to-be closed shard
```

DiannaHohensee committed:
account for idempotent shutdown API

DaveCTurner reviewed (changes requested):
I think we have to check for the shutdown type in all cases as in the patch I suggested, not just on the added-shutdown branch.

DiannaHohensee commented:
> I think we have to check for the shutdown type in all cases as in the patch I suggested, not just on the added-shutdown branch.

Hmm. In case shutdown moves from pausing type to RESTART type 🤔 Ooph.

DiannaHohensee committed:
check shutdown same for add and remove

DiannaHohensee committed:
rename and fix TestIssueLogging

DaveCTurner reviewed (approved):
LGTM
